Improve layers

MathemanFlo opened the pull request:
This PR tries to fix several issues related to layers and improves a few things.
1. New groups are not always available. Therefore it is better to grey out if not available.
2. Layer delete has several issues. It does not work well with group masks, the undo and redo system is broken and it was not possible to delete in all valid cases resp. there have been cases where deleting shouldn't have been available. This PR sorts them out and fixes  #989. Maybe my undo-redo approach is not the most efficient way but it works reliably as far as I can see.
3. Duplicate layers did not work well with group masks. Fixed these issues and refactored a bit.
4. canMergeDown could use a little bit of love to become more readable. Also it did not cover all cases I think.
5. Merging groups and merge down did not work reliably with group masks. Fixed these issues.
6. Mask invert could not be undone or redone. Implemented this
7. Creating a new group could not be undone or redone. Implemented this
8. Undo/redo for to fill mask or to paint mask had the more general to fill/paint layer labels. I improved this.

@luboslenco Please do not merge the PR yet. I'd like to sort out some more issues. The undo-redo for layer reordering is on the way. I only have to sort out a little issue that arises when the user collapses groups which is bad because this changes the behavior to add or not to add a layer to a group. At the moment you can expect that 

- #792 The non-linked issues there are fixed. The rearrange one is not pushed yet.
- #989 is fixed

1. Question: As you can see, I factored some methods out to make the code easier to read. Currently it is scattered around between TabLayers and Layers. Where should I put it? What about this distinction: LayerSlot.hx is the "dumb" one that does not know too much and does not know about undo-redo. The higher-order operations are implemented in Layers, know how to undo-redo and use LayerSlot. TabLayers will only be responsible for ui related stuff. 
2. I think the existing code is missing some `remapLayerPointers` isn't it? Whenever the indices change the pointers should be remapped, right? I guess there are some bugs to dig out. For instance what should happen if a layer disappears due to merging down.
3. `getMasks `is a bit strange for me. Why does it have the option to return all masks, including the group masks. Imho this is not the right behavior. I think one is probably never interested in all masks because both groups of masks do not play the same role. Second: I'd like to change the behavior to never return null but only to return an empty array if there is no mask. This would immediately remove all these null checks.
4. I think we have to think about the question what should happen if not all layers are shown because of the filter and what should happen if a group contains layers on varying meshes.

MathemanFlo commented:
Sorry for the compile error. Git had problems with not comitting a debugging method that was the last method of the class.

MathemanFlo commented:
It is necessary to delete layers on next init otherwise there are situations where this leads to problems. For example deleting a layer can lead to problems in some situations because the LayerSlot is becomes invalid.

MathemanFlo commented:
I will stop for today. Sorry that the number of commits escalated a bit. Happy to get your feedback 😃
Currently my plan is to do the following things
- [ ] Preserve layer/mask/group names in undo-redo
- [ ] Fix undo-redo for merge down, #1000, #1001, #1007,  #1089, 
- [ ] Fix undo-redo for merge group, #986, #987, 
- [ ] Fix undo-redo for layer drag and drop
- [ ] Fix undo-redo for duplicate, #985
- [ ] Fix undo-redo apply mask, #981, 
- [ ] Fix undo-redo for rearrange layers (half way done), #966, #980, #1016

MathemanFlo commented:
Any comments or feedback so far? My plan is to go on on friday.

luboslenco commented:
> If you look carefully TabLayers.hx and History.hx have "\n" line endings but "Layers.hx" has "\r\n" line endings. My commit changed it to "\n" line endings. What do you suggest?

Not sure what is going on, to me all of these look like CRLF after checkout on Windows. On push it should be converted to LF by git. Do you get inconsistent line endings between files after fresh clone?

MathemanFlo commented:
> > If you look carefully TabLayers.hx and History.hx have "\n" line endings but "Layers.hx" has "\r\n" line endings. My commit changed it to "\n" line endings. What do you suggest?
> 
> Not sure what is going on, to me all of these look like CRLF after checkout on Windows. On push it should be converted to LF by git. Do you get inconsistent line endings between files after fresh clone?

The line endings are consistent on checkout., everything is CRLF on Windows. But the files in the repository are not consistent. For example check https://raw.githubusercontent.com/armory3d/armorpaint/1ad67a2eb664ca24c826142dcc54bd32df4a5770/Sources/arm/Layers.hx
Download it and realize it is CRLF in repo. That's why I'm sure it is related to this one. I believe my git converts to LF and that's why the whole file is modified. The very same problem occured on two different maschines. Steps to reproduce
1. Modify Layers.hx
2. Partially add the changes to the commit. (In VSC it is Ctrl+K,Ctrl+Alt+S) or right click in the context menu while being in GIT-view. This is the evil step. If the whole file is added nothing weird happens.
3. Commit it
4. Push it

Fresh Example https://github.com/armory3d/armorpaint/commit/6a09ee0b65e557b52babe9e54290f56fffc06d70
Locally
![grafik](https://user-images.githubusercontent.com/28649121/157863705-b64a8035-3bed-4fee-a0e6-00266dd0be3e.png)
Sorry for bothering you...

luboslenco commented:
Great catch. Fixed now, hopefully I did not cause too much trouble.

MathemanFlo commented:
No it's ok. I rewrote the history once because I thought I made a weird mistake. Didn't change anything but still all lines changed without any visible reason. Thus line endings were a natural guess to investigate... Now rebased everything which worked well. Everything is fine now. Thank you for the quick fix. I'm really happy that the problem resolved so easily. Would not have wanted to recommit everything 😊
Any other comments so far?

MathemanFlo commented:
Oki 👍 Will go on the next days and fix all stuff. Then I propose to merge everything. In a second PR we cleanup the code and improve everything.

luboslenco commented:
Thanks again!

MathemanFlo commented:
Sorry for not continuing the last days but I was very bussy. Will continue my work in another PR. Probably it is better to merge more often and create smaller PRs.

luboslenco commented:
No prob at all.

> Probably it is better to merge more often and create smaller PRs.

👍👍
